presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java

```diff
@@ -373,8 +373,12 @@ public static long addFieldValueDate(ConnectorSession session, @SqlType("varchar
     @SqlType(StandardTypes.TIME)
     public static long addFieldValueTime(ConnectorSession session, @SqlType("varchar(x)") Slice unit, @SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.TIME) long time)
     {
-        ISOChronology chronology = getChronology(session.getTimeZoneKey());
-        return modulo24Hour(chronology, getTimeField(chronology, unit).add(time, toIntExact(value)));
+        if (session.isLegacyTimestamp()) {
+            ISOChronology chronology = getChronology(session.getTimeZoneKey());
+            return modulo24Hour(chronology, getTimeField(chronology, unit).add(time, toIntExact(value)));
+        }
+
+        return modulo24Hour(getTimeField(UTC_CHRONOLOGY, unit).add(time, toIntExact(value)));
```

Comment:
No. Stack representation of new TIME is a long value within `[0, 86_400_000)` range. However, legacy TIME stack representation (for a moment assuming a fix offset zone) is a long value within `[-offset, 86_400_000 - offset)`. 
`modulo24Hour` makes sure the return value is indeed within allowed range. If we returned value not in the allowed range, it would still print (or cast to varchar) correctly, but wouldn't pass equality check with corresponding value constructed via a literal (which in turn would be normalized).

Thus, before i added this check, i had tests saying `expected 3:04:05 but got 3:04:04`.

Reply:
Got it. Thanks